Log errors in ext pillars

jbertran opened the pull request:
**Component**: salt pillar

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: 

We currently only sporadically report errors in external pillar calculations, where we should do it consistently for ease of debug.

**Summary**:

* Add `_errors` reporting to nodes external pillar
* Add `_errors` reporting to endpoints external pillar
* Add `_errors` reporting to main external pillar and remove `assert`-based control flow management.

**Acceptance criteria**:

Pillars now report errors when you `get` them.
For example, 

```
salt-call pillar.get metalk8s.endpoints._errors
```

Should yield `None` in the nominal case.

```
salt-call pillar.get metalk8s.private._errors
```
Should complain when having moved the SA private key to an unexpected path.

---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #1018 

<!-- If you want to refer to an issue while not closing it, use:

See: #

-->

jbertran committed:
pillar: log to _errors in result dict for endpoints ext_pillar

Issue: #1018

jbertran committed:
pillar: log to _errors in result dict for nodes ext_pillar

Issue: #1018

jbertran committed:
pillar: report errors instead of raising AssertionErrors

Issue: #1018

jbertran committed:
pillar: log to _errors in result dict for main ext_pillar

Issue: #1018

bert-e commented:
# Hello jbertran,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/1018-log-errors-in-ext-pillars` and the
destination branch `development/2.0`.

Please resolve the conflict on **the feature branch** (`improvement/1018-log-errors-in-ext-pillars`).

```bash
 $ git fetch
 $ git checkout origin/improvement/1018-log-errors-in-ext-pillars
 $ git merge origin/development/2.0
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/1018-log-errors-in-ext-pillars
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/1018-log-errors-in-ext-pillars` into
`development/2.0`
* [w/2.1/improvement/1018-log-errors-in-ext-pillars](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/1018-log-errors-in-ext-pillars) will be merged into `development/2.1` 
* [w/2.2/improvement/1018-log-errors-in-ext-pillars](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/1018-log-errors-in-ext-pillars) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet

bert-e commented:
# History mismatch



Merge commit #90ed2aff145405eb4a90d1a456fbbbc2801b6cc4 on the integration branch
`w/2.1/improvement/1018-log-errors-in-ext-pillars` is merging a branch which is neither the current
branch `improvement/1018-log-errors-in-ext-pillars` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/1018-log-errors-in-ext-pillars` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented on salt/_pillar/metalk8s.py:
```diff
@@ -11,26 +11,58 @@
+        if key not in source_dict:
+            errors.append(errors_tplt.format(key, source_dict))
+
+    return errors
```
Maybe make sense to add these functions to a custom `utils` as it's could be usefull for all ext_pillar
https://docs.saltstack.com/en/latest/topics/utils/index.html

jbertran replied:
Yup, planned :slightly_smiling_face:

TeddyAndrieux commented on salt/_pillar/metalk8s.py:
```diff
@@ -100,12 +140,28 @@ def _load_iso_path(config_data):
+
+    if '_errors' in result['networks']:
+        result.setdefault('errors', []).append(
+            result['networks'].pop('_errors')
```
Hmm don't know if we really want this, because here if we do for example
`salt-call pillar.get metalk8s:networks` we will just get None.
For me erros should be really "local" to the section.
Maybe you can have errors for network but not for metalk8s so you don't want `error` directly at the root.

IMO what we want to be able to do it's in our state we get a value lets say `pillar.get('networks')`. We have `_errors` here ok fail with the error message(s). (But indeed have all error at the "root" of the pillar can also help ... Hmm maybe we should have both local + aggregation at the root of the pillar)

Maybe someone else has another opinion here (cc @NicolasT @gdemonet )

(Also `_` missing here before `errors` no ?)

jbertran replied:
For now this only goes as far as reproducing the existing `assert` control flow, so it seems strange in some places - indeed, it's better to have local errors, but the assertions used to make the whole thing crash violently before.

The local/global thing makes sense to me, I'll work it in unless I get any thoughts otherwise.

> (Also _ missing here before errors no ?)

Indeed!

jbertran committed:
pillar: consolidate utility functions into utils

Issue: #1018

gdemonet commented on salt/_pillar/utils.py:
```diff
@@ -0,0 +1,79 @@
+    errors = []
+    errors_tplt = "Expected presence of key '{}' in data: {}"
+
+    for key in keys:
```
Could have been a list comprehension as was done in `_assert_equals`

gdemonet commented on salt/_pillar/utils.py:
```diff
@@ -0,0 +1,79 @@
+    """
+    patch = source[key]
+    if patch.get('_errors'):
+        source.setdefault("_errors", []).extend(patch.get("_errors"))
```
So, leaving the `_errors` in the promoted `key`?

Let's say the `metalk8s_private` pillar fails for the SA private key. The pillar would look like:

```yaml
metalk8s:
  private:
    _errors:
      - 'SA private key not found'
    sa_private_key:
      _errors:
        - 'SA private key not found'
```

jbertran replied:
Yes. It's a little verbose for single keys, but when you're performing operations that can fail for **multiple** keys it's a little more interesting because we have all errors concerning a key at every level:

```yaml
networks:
  _errors:
    - ControlPlane not Found
metalk8s:
  ca:
    _errors:
      - /etc/kubernetes/admin.conf not found
  products:
    - foo/bar
  api_server:
    _errors:
      - API Server dead or broken pls fix
  _errors:
    - /etc/kubernetes/admin.conf not found
    - API Server dead or broken pls fix
_errors:
  - /etc/kubernetes/admin.conf not found
  - API Server dead or broken pls fix
  - /etc/kubernetes/admin.conf not found
```
This way, I get errors whether I ask for `pillar.get metalk8s` or `pillar.get metalk8s:ca:_errors`
We _could_ have some better logic with, say, a custom `pillar.get` that would recursively grab errors, but this is getting a little much as is :)

gdemonet commented on salt/_pillar/metalk8s.py:
```diff
@@ -3,6 +3,8 @@
 import salt.utils.files
 import salt.utils.yaml
 
+from . import utils
```
Maybe worth moving into the builtin `__utils__`

jbertran replied:
Yup, done.

gdemonet reviewed (commented):
A few comments regarding the utility functions, but mostly a good improvement! We would need to decide how we want a failing pillar to look like, given the use cases we want to implement:

- @TeddyAndrieux mentioned doing some `pillar.get metalk8s:somekey:_errors` for a user to debug a specific `ext_pillar`
- Maybe we need to use such keys in formulas to check if some pillar key is in error state or not

jbertran committed:
pillar: move utils to __utils__, no top-level _errors in ext_pillar

If we include a top-level `_errors` key in the ext_pillar response,
the bootstrap's initial stages fail because most resources are not yet
installed and a top-level `_errors` key causes Salt to infer that the call
should fail.

Issue: #1018

bert-e commented:
# History mismatch



Merge commit #b3824c0dd846a072e32ab1defe357edc4610e8c3 on the integration branch
`w/2.1/improvement/1018-log-errors-in-ext-pillars` is merging a branch which is neither the current
branch `improvement/1018-log-errors-in-ext-pillars` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/1018-log-errors-in-ext-pillars` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented on salt/_pillar/metalk8s_private.py:
```diff
@@ -1,6 +1,7 @@
 import os.path
 import logging
 
+import pillar_utils
```
For me you don't need this and you should not use this.

You should call utils using `__utils__` like
`__utils__['pillar_utils.update_with_errors'](blabla)`

jbertran committed:
bootstrap: sync __utils__, sync everything after minion key accept

Issue: #1018

TeddyAndrieux commented on salt/_pillar/metalk8s.py:
```diff
@@ -3,6 +3,8 @@
 import salt.utils.files
 import salt.utils.yaml
 
+import pillar_utils
```
This is no longer needed

jbertran replied:
Oops indeed.

TeddyAndrieux commented on salt/_pillar/metalk8s.py:
```diff
@@ -11,26 +13,42 @@
 DEFAULT_SERVICE_NETWORK = '10.96.0.0/12'
 
 
+
```
`salt/_pillar/metalk8s.py:16:1: E303 too many blank lines (3)`

TeddyAndrieux commented on scripts/bootstrap.sh.in:
```diff
@@ -296,6 +300,11 @@ orchestrate_bootstrap() {
             saltenv=metalk8s-@@VERSION \
             pillar="${pillar[*]}"
 
+    run "Refresh pillar data on bootstrap minion" \
```
To be consistent
```suggestion
    run "Refreshing pillar data on bootstrap minion" \
```

jbertran commented:
@bert-e approve

bert-e commented:
# History mismatch



Merge commit #b3824c0dd846a072e32ab1defe357edc4610e8c3 on the integration branch
`w/2.1/improvement/1018-log-errors-in-ext-pillars` is merging a branch which is neither the current
branch `improvement/1018-log-errors-in-ext-pillars` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/1018-log-errors-in-ext-pillars` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

jbertran commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/1018-log-errors-in-ext-pillars` into
`development/2.0`
* [w/2.1/improvement/1018-log-errors-in-ext-pillars](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/1018-log-errors-in-ext-pillars) will be merged into `development/2.1` 
* [w/2.2/improvement/1018-log-errors-in-ext-pillars](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/1018-log-errors-in-ext-pillars) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye jbertran.
